app/helpers/messages_helper.rb

```diff
@@ -1,2 +1,71 @@
 module MessagesHelper
+  def format(text)
+    text = html_escape(text)
+
+    if text.include?('`')
+      text = paired_single_backticks(text)
+      text = paired_triple_backticks(text)
+      text = copy_button(text)
+    end
+
+    text.html_safe
+  end
+
+  def paired_single_backticks(text)
+    text.gsub(/(?<!`)(`)(?=[^`]*`)([^`]+)`(?!`)/,
+      '<code><span class="hidden">`</span>\2<span class="hidden">`</span></code>')
+  end
+
+  def paired_triple_backticks(text)
+    text.gsub(/(?<!`)(```)([a-zA-Z0-9]*)(?:\n)?((?:\n?.*?)*)```(?!`)(?:\n)?/m) do
+      language = $2
```

Comment:
I just got a response from chatgpt that looks like this:

3. **Canopy by Hilton San Francisco SoMa** - A modern hotel with flexible meeting spaces that can be adjusted for any type of gathering. Its location in the SoMa district is also ideal for exploring the city during downtime.

It occurred to me that chatgpt just replies in markdown. Here I went and prioritize the code markdown but maybe I did it in such a way that will lead me to generate gsub for all the key markdown tags. That won’t be too hard, but if I would have stepped back and just said: how do I parse the whole response as markdown, it may have led me to a different solution. I’m inclined to ship what I did, anyway, but before starting to parse other tags I should investigate ruby markdown parsers.

Reply:
What does Claude return?
Maybe check what Brakeman has to say about these regex (which will become a default gem in Rails 8)

Since you have a copy of Campfire, maybe check how they handle it. I know they're loading a bunch of language files.

Else, may be enough for now but leveraging some parsers look like a more robust approach on the long run.
Here is an example https://www.driftingruby.com/episodes/streaming-llm-responses

Reply:
This episode was a helpful link. It led me down a rabbit hole and ultimately to Redcarpet, which is a nice markdown parser that's pretty robust. I switched our my regex approach to use this. The bonus is *all* markdown now gets parsed. It's a nice win.